src/java/org/pantsbuild/tools/junit/impl/ShutdownListener.java

```diff
@@ -3,71 +3,88 @@
 
 package org.pantsbuild.tools.junit.impl;
 
-import java.io.PrintStream;
+import java.util.concurrent.ConcurrentHashMap;
 import org.junit.runner.Description;
 import org.junit.runner.Result;
 import org.junit.runner.notification.Failure;
 import org.junit.runner.notification.RunListener;
 
 /**
- * A listener that keeps track of the current test state with its own result class so it can print
+ * A listener that keeps track of the current test state with its own result class so it can record
  * the state of tests being run if there is unexpected exit during the tests.
  */
-public class ShutdownListener extends ConsoleListener {
+public class ShutdownListener extends RunListener {
   private final Result result = new Result();
   private final RunListener resultListener = result.createListener();
-  private Description currentTestDescription;
+  private ConcurrentHashMap.KeySetView<Description, Boolean> currentDescriptions =
```

Comment:
It would be good to describe the "protocol" with which this map is updated. It looks like failures don't remove from this map (`testFailure`), but finishing cleanly (`testFinished`) does? It's not clear why.

Reply:
RunListener says testFinished is "Called when an atomic test has finished, whether the test succeeds or fails" so the protocol is to track when it starts, and untrack when it finishes. Should I add that protocol as a comment here?

Reply:
Yes please.